pkg/kn/commands/source/human_readable_flags.go

```diff
@@ -85,3 +101,148 @@ func printSourceTypesList(sourceTypesList *unstructured.UnstructuredList, option
 	}
 	return rows, nil
 }
+
+// printSource populates a single row of source list table
+func printSource(source unstructured.Unstructured, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
+	name := source.GetName()
+	sourceType := source.GetKind()
+	sourceTypeName := getSourceTypeName(source)
+	sink := findSink(source)
+	ready := isReady(source)
+
+	row := metav1beta1.TableRow{
+		Object: runtime.RawExtension{Object: &source},
+	}
+
+	if options.AllNamespaces {
+		row.Cells = append(row.Cells, source.GetNamespace())
+	}
+
+	row.Cells = append(row.Cells, name, sourceType, sourceTypeName, sink, ready)
+	return []metav1beta1.TableRow{row}, nil
+}
+
+// printSourceList populates the source list table rows
+func printSourceList(sourceList *unstructured.UnstructuredList, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
```

Comment:
So you could use a `sources []*Unstructured` as argument instead of `sourceList` here, too, as you only iterate of the list `items`, right ?

Reply:
can't pass the slice as this function needs an object because:

https://github.com/knative/client/blob/a44833c849547a13048426b421b711267412db25/pkg/printers/interface.go#L26-L29

```
--- a/pkg/kn/commands/source/human_readable_flags.go
+++ b/pkg/kn/commands/source/human_readable_flags.go
@@ -123,14 +123,14 @@ func printSource(source unstructured.Unstructured, options printers.PrintOptions
 }
 
 // printSourceList populates the source list table rows
-func printSourceList(sourceList *unstructured.UnstructuredList, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
-	rows := make([]metav1beta1.TableRow, 0, len(sourceList.Items))
+func printSourceList(sourceList []unstructured.Unstructured, options printers.PrintOptions) ([]metav1beta1.TableRow, error) {
+	rows := make([]metav1beta1.TableRow, 0, len(sourceList))
 
-	sort.SliceStable(sourceList.Items, func(i, j int) bool {
-		return sourceList.Items[i].GetName() < sourceList.Items[j].GetName()
+	sort.SliceStable(sourceList, func(i, j int) bool {
+		return sourceList[i].GetName() < sourceList[j].GetName()
 	})
-	for _, item := range sourceList.Items {
-		row, err := printSource(item, options)
+	for _, source := range sourceList {
+		row, err := printSource(source, options)
 		if err != nil {
 			return nil, err
 		}
diff --git a/pkg/kn/commands/source/list.go b/pkg/kn/commands/source/list.go
index 2f157e17..c37dd0b3 100644
--- a/pkg/kn/commands/source/list.go
+++ b/pkg/kn/commands/source/list.go
@@ -65,7 +65,7 @@ func NewListCommand(p *commands.KnParams) *cobra.Command {
 			if err != nil {
 				return nil
 			}
-			err = printer.PrintObj(sourceList, cmd.OutOrStdout())
+			err = printer.PrintObj(sourceList.Items, cmd.OutOrStdout())
 			if err != nil {
 				return err
 			}
➜  client git:(pr/source-list) ✗ ./hack/build.sh -f
🚧 Compile
# knative.dev/client/pkg/kn/commands/source
pkg/kn/commands/source/list.go:68:37: cannot use sourceList.Items (type []unstructured.Unstructured) as type "k8s.io/apimachinery/pkg/runtime".Object in argument to printer.PrintObj:
	[]unstructured.Unstructured does not implement "k8s.io/apimachinery/pkg/runtime".Object (missing DeepCopyObject method)

```

Reply:
Yes, I see. The funny thing is, that nothing really requires it to be a `runtime.Object` in the printed code. Its only in the print function which might extract this. So I wonder whether we just implement runtime.Object on the slice (with no-ops), like in

```golang
type sourceList []*Unstructured

func (s sourceList) GetObjectKind() {
    return EmptyObjectKind
}

func (s sourceList) DeepCopy(...) {
    ....
}
```
(you get the idea)

Regardless what, maybe just use the UnstructuredList for now, but set the the kind to `EmptyObjectKind`.